File: python/eventgraph/trust.py
"""Trust model — asymmetric, non-transitive, decaying trust with domain scoping."""

from __future__ import annotations

import threading
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable

from .event import Event
from .types import ActorID, DomainScope, EventID, Score, Weight


# ── TrustMetrics ─────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class TrustMetrics:
    """Immutable snapshot of an actor's trust state."""

    _actor: ActorID
    _overall: Score
    _by_domain: dict[str, Score]
    _confidence: Score
    _trend: Weight
    _evidence: list[EventID]
    _last_updated: int
    _decay_rate: Score

    def __init__(
        self,
        actor: ActorID,
        overall: Score,
        by_domain: dict[str, Score],
        confidence: Score,
        trend: Weight,
        evidence: list[EventID],
        last_updated: int,
        decay_rate: Score,
    ) -> None:
        object.__setattr__(self, "_actor", actor)
        object.__setattr__(self, "_overall", overall)
        object.__setattr__(self, "_by_domain", dict(by_domain))
        object.__setattr__(self, "_confidence", confidence)
        object.__setattr__(self, "_trend", trend)
        object.__setattr__(self, "_evidence", list(evidence))
        object.__setattr__(self, "_last_updated", last_updated)
        object.__setattr__(self, "_decay_rate", decay_rate)

    @property
    def actor(self) -> ActorID:
        return self._actor

    @property
    def overall(self) -> Score:
        return self._overall

    @property
    def by_domain(self) -> dict[str, Score]:
        return dict(self._by_domain)

    @property
    def confidence(self) -> Score:
        return self._confidence

    @property
    def trend(self) -> Weight:
        return self._trend

    @property
    def evidence(self) -> list[EventID]:
        return list(self._evidence)

    @property
    def last_updated(self) -> int:
        return self._last_updated

    @property
    def decay_rate(self) -> Score:
        return self._decay_rate


# ── TrustConfig ──────────────────────────────────────────────────────────


@dataclass(frozen=True, slots=True)
class TrustConfig:
    """Configuration for the default trust model."""

    initial_trust: Score = field(default_factory=lambda: Score(0.0))
    decay_rate: Score = field(default_factory=lambda: Score(0.01))
    max_adjustment: Weight = field(default_factory=lambda: Weight(0.1))
    observed_event_delta: float = 0.01
    trend_decay_rate: float = 0.01


# ── TrustModel Protocol ─────────────────────────────────────────────────


@runtime_checkable
class TrustModel(Protocol):
    """Protocol for trust calculation, update, and decay."""

    def score(self, actor: ActorID) -> TrustMetrics: ...

    def score_in_domain(self, actor: ActorID, domain: DomainScope) -> TrustMetrics: ...

    def update(self, actor: ActorID, evidence: Event) -> TrustMetrics: ...

    def update_between(
        self, from_actor: ActorID, to_actor: ActorID, evidence: Event
    ) -> TrustMetrics: ...

    def decay(self, actor: ActorID, elapsed_seconds: float) -> TrustMetrics: ...

    def between(self, from_actor: ActorID, to_actor: ActorID) -> TrustMetrics: ...


# ── Internal state ───────────────────────────────────────────────────────


class _TrustState:
    """Mutable internal trust state for a single actor or directed pair."""

    __slots__ = ("score", "by_domain", "evidence", "last_updated", "trend")

    def __init__(self, initial_score: float, last_updated: int) -> None:
        self.score: float = initial_score
        self.by_domain: dict[str, float] = {}
        self.evidence: list[EventID] = []
        self.last_updated: int = last_updated
        self.trend: float = 0.0


# ── DefaultTrustModel ────────────────────────────────────────────────────


def _now_nanos() -> int:
    import time

    return int(time.time() * 1_000_000_000)


class DefaultTrustModel:
    """Thread-safe trust model with linear decay and equal weighting.

    Implements the TrustModel protocol.
    """

    def __init__(self, config: TrustConfig | None = None) -> None:
        self._config = config or TrustConfig()
        self._lock = threading.Lock()
        self._scores: dict[str, _TrustState] = {}  # actor_id.value -> state
        self._directed: dict[tuple[str, str], _TrustState] = {}  # (from, to) -> state

    # ── Internal helpers ─────────────────────────────────────────────────

    def _get_or_create(self, actor_id: ActorID) -> _TrustState:
        """Get or create trust state. Caller must hold self._lock."""
        key = actor_id.value
        if key not in self._scores:
            self._scores[key] = _TrustState(
                self._config.initial_trust.value, _now_nanos()
            )
        return self._scores[key]

    def _get_or_default(self, actor_id: ActorID) -> _TrustState:
        """Get trust state or return a default without mutating storage.
        Caller must hold self._lock.
        """
        key = actor_id.value
        if key in self._scores:
            return self._scores[key]
        return _TrustState(self._config.initial_trust.value, _now_nanos())

    def _build_metrics(self, actor_id: ActorID, state: _TrustState) -> TrustMetrics:
        evidence_count = len(state.evidence)
        confidence = min(1.0, evidence_count / 50.0)
        by_domain = {k: Score(v) for k, v in state.by_domain.items()}
        return TrustMetrics(
            actor=actor_id,
            overall=Score(state.score),
            by_domain=by_domain,
            confidence=Score(confidence),
            trend=Weight(state.trend),
            evidence=list(state.evidence),
            last_updated=state.last_updated,
            decay_rate=self._config.decay_rate,
        )

    def _build_domain_metrics(
        self, actor_id: ActorID, state: _TrustState, domain_score: float
    ) -> TrustMetrics:
        evidence_count = len(state.evidence)
        confidence = min(1.0, evidence_count / 50.0)
        by_domain = {k: Score(v) for k, v in state.by_domain.items()}
        return TrustMetrics(
            actor=actor_id,
            overall=Score(domain_score),
            by_domain=by_domain,
            confidence=Score(confidence),
            trend=Weight(state.trend),
            evidence=list(state.evidence),
            last_updated=state.last_updated,
            decay_rate=self._config.decay_rate,
        )

    def _extract_delta(self, evidence: Event, current_score: float) -> float:
        """Extract trust delta from evidence event content."""
        content = evidence.content
        if isinstance(content, dict) and "current" in content:
            return content["current"] - current_score
        return self._config.observed_event_delta

    def _apply_update(self, state: _TrustState, evidence: Event) -> None:
        """Apply an evidence event to a trust state. Caller must hold self._lock."""
        # Deduplicate
        for eid in state.evidence:
            if eid.value == evidence.id.value:
                return

        delta = self._extract_delta(evidence, state.score)

        # Clamp to max_adjustment
        max_adj = self._config.max_adjustment.value
        if delta > max_adj:
            delta = max_adj
        if delta < -max_adj:
            delta = -max_adj

        # Apply delta, clamp to [0, 1]
        new_score = max(0.0, min(1.0, state.score + delta))
        state.score = new_score

        # Update domain-specific score if evidence carries domain info
        content = evidence.content
        if isinstance(content, dict) and "domain" in content:
            domain_key = content["domain"]
            if domain_key in state.by_domain:
                domain_score = max(0.0, min(1.0, state.by_domain[domain_key] + delta))
            else:
                domain_score = max(
                    0.0, min(1.0, self._config.initial_trust.value + delta)
                )
            state.by_domain[domain_key] = domain_score

        # Update trend
        if delta > 0:
            state.trend = min(1.0, state.trend + 0.1)
        elif delta < 0:
            state.trend = max(-1.0, state.trend - 0.1)

        # Track evidence (cap at 100)
        state.evidence.append(evidence.id)
        if len(state.evidence) > 100:
            state.evidence = state.evidence[len(state.evidence) - 100 :]

        state.last_updated = _now_nanos()

    def _decay_state(
        self, state: _TrustState, decay_amount: float, trend_decay: float
    ) -> None:
        """Apply linear decay to a trust state. Caller must hold self._lock."""
        state.score = max(0.0, state.score - decay_amount)

        for domain_key in state.by_domain:
            state.by_domain[domain_key] = max(
                0.0, state.by_domain[domain_key] - decay_amount
            )

        if state.trend > 0:
            state.trend = max(0.0, state.trend - trend_decay)
        elif state.trend < 0:
            state.trend = min(0.0, state.trend + trend_decay)

        state.last_updated = _now_nanos()

    # ── Public API (TrustModel protocol) ─────────────────────────────────

    def score(self, actor: ActorID) -> TrustMetrics:
        """Return current trust metrics for an actor."""
        with self._lock:
            state = self._get_or_default(actor)
            return self._build_metrics(actor, state)

    def score_in_domain(self, actor: ActorID, domain: DomainScope) -> TrustMetrics:
        """Return domain-specific trust, falling back to global with halved confidence."""
        with self._lock:
            state = self._get_or_default(actor)

            if domain.value in state.by_domain:
                return self._build_domain_metrics(
                    actor, state, state.by_domain[domain.value]
                )

            # Fall back to global score with halved confidence
            evidence_count = len(state.evidence)
            global_confidence = min(1.0, evidence_count / 50.0)
            by_domain = {k: Score(v) for k, v in state.by_domain.items()}
            return TrustMetrics(
                actor=actor,
                overall=Score(state.score),
                by_domain=by_domain,
                confidence=Score(global_confidence * 0.5),
                trend=Weight(state.trend),
                evidence=list(state.evidence),
                last_updated=state.last_updated,
                decay_rate=self._config.decay_rate,
            )

    def update(self, actor: ActorID, evidence: Event) -> TrustMetrics:
        """Update trust based on evidence event."""
        with self._lock:
            state = self._get_or_create(actor)

            # Check deduplication before applying
            for eid in state.evidence:
                if eid.value == evidence.id.value:
                    return self._build_metrics(actor, state)

            self._apply_update(state, evidence)
            return self._build_metrics(actor, state)

    def update_between(
        self, from_actor: ActorID, to_actor: ActorID, evidence: Event
    ) -> TrustMetrics:
        """Update directional trust from one actor toward another."""
        with self._lock:
            key = (from_actor.value, to_actor.value)
            if key not in self._directed:
                self._directed[key] = _TrustState(
                    self._config.initial_trust.value, _now_nanos()
                )
            state = self._directed[key]

            # Check deduplication before applying
            for eid in state.evidence:
                if eid.value == evidence.id.value:
                    return self._build_metrics(to_actor, state)

            self._apply_update(state, evidence)
            return self._build_metrics(to_actor, state)

    def decay(self, actor: ActorID, elapsed_seconds: float) -> TrustMetrics:
        """Decay trust over elapsed time. Also decays directed trust where actor is from."""
        with self._lock:
            if elapsed_seconds <= 0:
                state = self._get_or_default(actor)
                return self._build_metrics(actor, state)

            days = elapsed_seconds / 86400.0
            decay_amount = self._config.decay_rate.value * days
            trend_decay = self._config.trend_decay_rate * days

            # Decay undirected trust
            key = actor.value
            has_state = key in self._scores
            if has_state:
                self._decay_state(self._scores[key], decay_amount, trend_decay)

            # Decay directed trust where actor is the from side
            for dkey, dstate in self._directed.items():
                if dkey[0] == actor.value:
                    self._decay_state(dstate, decay_amount, trend_decay)

            if not has_state:
                defaults = _TrustState(self._config.initial_trust.value, _now_nanos())
                return self._build_metrics(actor, defaults)

            return self._build_metrics(actor, self._scores[key])

    def between(self, from_actor: ActorID, to_actor: ActorID) -> TrustMetrics:
        """Return directed trust from one actor to another."""
        with self._lock:
            key = (from_actor.value, to_actor.value)
            state = self._directed.get(key)
            if state is None:
                return TrustMetrics(
                    actor=to_actor,
                    overall=self._config.initial_trust,
                    by_domain={},
                    confidence=Score(0.0),
                    trend=Weight(0.0),
                    evidence=[],
                    last_updated=_now_nanos(),
                    decay_rate=self._config.decay_rate,
                )
            return self._build_metrics(to_actor, state)

File: dotnet/src/EventGraph/Trust.cs
namespace EventGraph;

// ── TrustMetrics ────────────────────────────────────────────────────────

/// <summary>Immutable trust metrics snapshot for an actor.</summary>
public sealed class TrustMetrics
{
    public ActorId Actor { get; }
    public Score Overall { get; }
    private readonly Dictionary<string, Score> _byDomain;
    public IReadOnlyDictionary<string, Score> ByDomain => new Dictionary<string, Score>(_byDomain);
    public Score Confidence { get; }
    public Weight Trend { get; }
    private readonly List<EventId> _evidence;
    public IReadOnlyList<EventId> Evidence => _evidence.AsReadOnly();
    public long LastUpdatedNanos { get; }
    public Score DecayRate { get; }

    public TrustMetrics(
        ActorId actor,
        Score overall,
        IReadOnlyDictionary<string, Score>? byDomain,
        Score confidence,
        Weight trend,
        IReadOnlyList<EventId>? evidence,
        long lastUpdatedNanos,
        Score decayRate)
    {
        Actor = actor;
        Overall = overall;
        _byDomain = byDomain is not null
            ? new Dictionary<string, Score>(byDomain)
            : new Dictionary<string, Score>();
        Confidence = confidence;
        Trend = trend;
        _evidence = evidence is not null
            ? new List<EventId>(evidence)
            : new List<EventId>();
        LastUpdatedNanos = lastUpdatedNanos;
        DecayRate = decayRate;
    }
}

// ── TrustConfig ─────────────────────────────────────────────────────────

/// <summary>Configuration for the default trust model.</summary>
public record TrustConfig
{
    public Score InitialTrust { get; init; } = new(0.0);
    public Score DecayRate { get; init; } = new(0.01);
    public Weight MaxAdjustment { get; init; } = new(0.1);
    public double ObservedEventDelta { get; init; } = 0.01;
    public double TrendDecayRate { get; init; } = 0.01;
}

// ── ITrustModel ─────────────────────────────────────────────────────────

/// <summary>Calculates, updates, and decays trust.</summary>
public interface ITrustModel
{
    /// <summary>Get current trust metrics for an actor.</summary>
    TrustMetrics Score(Actor actor);

    /// <summary>Get trust metrics for an actor in a specific domain.</summary>
    TrustMetrics ScoreInDomain(Actor actor, DomainScope domain);

    /// <summary>Update trust for an actor based on evidence.</summary>
    TrustMetrics Update(Actor actor, Event evidence);

    /// <summary>Update directional trust from one actor toward another.</summary>
    TrustMetrics UpdateBetween(Actor from, Actor to, Event evidence);

    /// <summary>Apply time-based trust decay for an actor.</summary>
    TrustMetrics Decay(Actor actor, TimeSpan elapsed);

    /// <summary>Get directional trust from one actor toward another.</summary>
    TrustMetrics Between(Actor from, Actor to);
}

// ── DefaultTrustModel ───────────────────────────────────────────────────

/// <summary>Thread-safe default trust model with linear decay and equal weighting.</summary>
public sealed class DefaultTrustModel : ITrustModel
{
    private readonly TrustConfig _config;
    private readonly object _lock = new();
    private readonly Dictionary<string, TrustState> _scores = new();
    private readonly Dictionary<(string from, string to), TrustState> _directed = new();

    private sealed class TrustState
    {
        public Score Score { get; set; }
        public Dictionary<string, Score> ByDomain { get; } = new();
        public List<EventId> Evidence { get; } = new();
        public long LastUpdatedNanos { get; set; }
        public Weight Trend { get; set; }
    }

    public DefaultTrustModel() : this(new TrustConfig()) { }

    public DefaultTrustModel(TrustConfig config)
    {
        _config = config;
    }

    private static long NowNanos() =>
        DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() * 1_000_000;

    private TrustState GetOrCreate(string actorId)
    {
        if (_scores.TryGetValue(actorId, out var state))
            return state;

        state = new TrustState
        {
            Score = _config.InitialTrust,
            LastUpdatedNanos = NowNanos(),
            Trend = new Weight(0.0),
        };
        _scores[actorId] = state;
        return state;
    }

    private TrustState GetOrDefault(string actorId)
    {
        if (_scores.TryGetValue(actorId, out var state))
            return state;

        // Return a default without storing it
        return new TrustState
        {
            Score = _config.InitialTrust,
            LastUpdatedNanos = NowNanos(),
            Trend = new Weight(0.0),
        };
    }

    public TrustMetrics Score(Actor actor)
    {
        lock (_lock)
        {
            var state = GetOrDefault(actor.Id.Value);
            return BuildMetrics(actor.Id, state);
        }
    }

    public TrustMetrics ScoreInDomain(Actor actor, DomainScope domain)
    {
        lock (_lock)
        {
            var state = GetOrDefault(actor.Id.Value);

            // If domain-specific score exists, return metrics with that score
            if (state.ByDomain.TryGetValue(domain.Value, out var domainScore))
                return BuildDomainMetrics(actor.Id, state, domainScore);

            // Fall back to global score with halved confidence
            var evidenceCount = state.Evidence.Count;
            var globalConfidence = Math.Min(1.0, evidenceCount / 50.0);
            return new TrustMetrics(
                actor.Id,
                state.Score,
                state.ByDomain,
                new Score(globalConfidence * 0.5),
                state.Trend,
                state.Evidence,
                state.LastUpdatedNanos,
                _config.DecayRate);
        }
    }

    public TrustMetrics Update(Actor actor, Event evidence)
    {
        lock (_lock)
        {
            var state = GetOrCreate(actor.Id.Value);

            // Deduplicate
            foreach (var id in state.Evidence)
            {
                if (id == evidence.Id)
                    return BuildMetrics(actor.Id, state);
            }

            var delta = ExtractDeltaForScore(evidence, state.Score.Value);

            // Clamp to MaxAdjustment
            var maxAdj = _config.MaxAdjustment.Value;
            delta = Math.Clamp(delta, -maxAdj, maxAdj);

            // Apply delta, clamp to [0, 1]
            var newScore = Math.Clamp(state.Score.Value + delta, 0.0, 1.0);
            state.Score = new Score(newScore);

            // Update domain-specific score if evidence carries domain info
            UpdateDomainFromEvidence(state, evidence, delta);

            // Update trend
            if (delta > 0)
                state.Trend = new Weight(Math.Min(1.0, state.Trend.Value + 0.1));
            else if (delta < 0)
                state.Trend = new Weight(Math.Max(-1.0, state.Trend.Value - 0.1));

            // Track evidence, cap at 100
            state.Evidence.Add(evidence.Id);
            if (state.Evidence.Count > 100)
                state.Evidence.RemoveRange(0, state.Evidence.Count - 100);

            state.LastUpdatedNanos = NowNanos();

            return BuildMetrics(actor.Id, state);
        }
    }

    public TrustMetrics UpdateBetween(Actor from, Actor to, Event evidence)
    {
        lock (_lock)
        {
            var key = (from: from.Id.Value, to: to.Id.Value);
            if (!_directed.TryGetValue(key, out var state))
            {
                state = new TrustState
                {
                    Score = _config.InitialTrust,
                    LastUpdatedNanos = NowNanos(),
                    Trend = new Weight(0.0),
                };
                _directed[key] = state;
            }

            // Deduplicate
            foreach (var id in state.Evidence)
            {
                if (id == evidence.Id)
                    return BuildMetrics(to.Id, state);
            }

            var delta = ExtractDeltaForScore(evidence, state.Score.Value);

            var maxAdj = _config.MaxAdjustment.Value;
            delta = Math.Clamp(delta, -maxAdj, maxAdj);

            var newScore = Math.Clamp(state.Score.Value + delta, 0.0, 1.0);
            state.Score = new Score(newScore);

            UpdateDomainFromEvidence(state, evidence, delta);

            if (delta > 0)
                state.Trend = new Weight(Math.Min(1.0, state.Trend.Value + 0.1));
            else if (delta < 0)
                state.Trend = new Weight(Math.Max(-1.0, state.Trend.Value - 0.1));

            state.Evidence.Add(evidence.Id);
            if (state.Evidence.Count > 100)
                state.Evidence.RemoveRange(0, state.Evidence.Count - 100);

            state.LastUpdatedNanos = NowNanos();

            return BuildMetrics(to.Id, state);
        }
    }

    public TrustMetrics Decay(Actor actor, TimeSpan elapsed)
    {
        lock (_lock)
        {
            // Guard against negative durations
            if (elapsed <= TimeSpan.Zero)
            {
                var defaultState = GetOrDefault(actor.Id.Value);
                return BuildMetrics(actor.Id, defaultState);
            }

            var days = elapsed.TotalDays;
            var decayAmount = _config.DecayRate.Value * days;
            var trendDecay = _config.TrendDecayRate * days;

            // Decay undirected trust
            var hasUndirected = _scores.TryGetValue(actor.Id.Value, out var state);
            if (hasUndirected)
                DecayState(state!, decayAmount, trendDecay);

            // Decay directed trust where actor is "from"
            foreach (var kvp in _directed)
            {
                if (kvp.Key.from == actor.Id.Value)
                    DecayState(kvp.Value, decayAmount, trendDecay);
            }

            if (!hasUndirected)
            {
                return new TrustMetrics(
                    actor.Id,
                    _config.InitialTrust,
                    null,
                    new Score(0.0),
                    new Weight(0.0),
                    null,
                    NowNanos(),
                    _config.DecayRate);
            }

            return BuildMetrics(actor.Id, state!);
        }
    }

    public TrustMetrics Between(Actor from, Actor to)
    {
        lock (_lock)
        {
            var key = (from: from.Id.Value, to: to.Id.Value);
            if (_directed.TryGetValue(key, out var state))
                return BuildMetrics(to.Id, state);

            // No direct trust relationship
            return new TrustMetrics(
                to.Id,
                _config.InitialTrust,
                null,
                new Score(0.0),
                new Weight(0.0),
                null,
                NowNanos(),
                _config.DecayRate);
        }
    }

    // ── Private helpers ────────────────────────────────────────────────

    private double ExtractDeltaForScore(Event ev, double currentScore)
    {
        var content = ev.Content;
        if (content.TryGetValue("current", out var currentObj) && currentObj is not null)
        {
            var current = Convert.ToDouble(currentObj);
            return current - currentScore;
        }
        return _config.ObservedEventDelta;
    }

    private void UpdateDomainFromEvidence(TrustState state, Event evidence, double delta)
    {
        var content = evidence.Content;
        if (content.TryGetValue("domain", out var domainObj) && domainObj is string domainStr)
        {
            double domainScore;
            if (state.ByDomain.TryGetValue(domainStr, out var existing))
                domainScore = Math.Clamp(existing.Value + delta, 0.0, 1.0);
            else
                domainScore = Math.Clamp(_config.InitialTrust.Value + delta, 0.0, 1.0);

            state.ByDomain[domainStr] = new Score(domainScore);
        }
    }

    private static void DecayState(TrustState state, double decayAmount, double trendDecay)
    {
        state.Score = new Score(Math.Max(0.0, state.Score.Value - decayAmount));

        foreach (var domain in state.ByDomain.Keys.ToList())
        {
            state.ByDomain[domain] = new Score(Math.Max(0.0, state.ByDomain[domain].Value - decayAmount));
        }

        if (state.Trend.Value > 0)
            state.Trend = new Weight(Math.Max(0.0, state.Trend.Value - trendDecay));
        else if (state.Trend.Value < 0)
            state.Trend = new Weight(Math.Min(0.0, state.Trend.Value + trendDecay));

        state.LastUpdatedNanos = NowNanos();
    }

    private TrustMetrics BuildMetrics(ActorId actorId, TrustState state)
    {
        var evidenceCount = state.Evidence.Count;
        var confidence = Math.Min(1.0, evidenceCount / 50.0);

        return new TrustMetrics(
            actorId,
            state.Score,
            state.ByDomain,
            new Score(confidence),
            state.Trend,
            state.Evidence,
            state.LastUpdatedNanos,
            _config.DecayRate);
    }

    private TrustMetrics BuildDomainMetrics(ActorId actorId, TrustState state, Score domainScore)
    {
        var evidenceCount = state.Evidence.Count;
        var confidence = Math.Min(1.0, evidenceCount / 50.0);

        return new TrustMetrics(
            actorId,
            domainScore,
            state.ByDomain,
            new Score(confidence),
            state.Trend,
            state.Evidence,
            state.LastUpdatedNanos,
            _config.DecayRate);
    }
}
